back/boxtribute_server/models/crud/box_transfer/agreement.py

```diff
@@ -0,0 +1,183 @@
+            for b in Base.select(Base.id).where(Base.organisation_id == organisation_id)
+        ]
+        invalid_base_ids = [i for i in base_ids if i not in organisation_base_ids]
+        if invalid_base_ids:
+            raise InvalidTransferAgreementBase(
+                expected_base_ids=organisation_base_ids,
+                base_id=invalid_base_ids[0],
+            )
+
+
+def create_transfer_agreement(
+    *,
+    target_organisation_id,
+    type,
+    source_base_ids=None,
+    target_base_ids=None,
+    valid_from=None,
+    valid_until=None,
+    timezone=None,
+    user,
+):
+    """Insert information for a new TransferAgreement in the database. Update
+    TransferAgreementDetail model with given source/target base information. By default,
+    the agreement is established between all bases of both organisations (indicated by
+    NULL for the Detail.source/target_base field).
+    Convert optional local dates into UTC datetimes using timezone information.
+    Raise an InvalidTransferAgreementOrganisation exception if the current user's
+    organisation is identical to the target organisation.
+    Raise an InvalidTransferAgreementBase expection if any specified source/target base
+    is not part of the source/target organisation.
+    """
+    source_organisation_id = user["organisation_id"]
+    if source_organisation_id == target_organisation_id:
+        raise InvalidTransferAgreementOrganisation()
+
+    with db.database.atomic():
+        if valid_from is not None or valid_until is not None:
+            tzinfo = tz.gettz(timezone)
+            # Insert time information such that start/end is at midnight
+            if valid_from is not None:
+                valid_from = datetime.combine(
+                    valid_from, time(), tzinfo=tzinfo
+                ).astimezone(dtimezone.utc)
+            if valid_until is not None:
+                valid_until = datetime.combine(
+                    valid_until, time(23, 59, 59), tzinfo=tzinfo
+                ).astimezone(dtimezone.utc)
+
+        transfer_agreement = TransferAgreement.create(
+            source_organisation=source_organisation_id,
+            target_organisation=target_organisation_id,
+            type=type,
+            valid_from=valid_from or utcnow(),
+            valid_until=valid_until,
+            requested_by=user["id"],
+        )
+
+        # In GraphQL input, base IDs can be omitted, or explicitly be null.
+        # Avoid duplicate base IDs by creating sets
+        source_base_ids = set(source_base_ids or [None])
+        target_base_ids = set(target_base_ids or [None])
+
+        _validate_bases_as_part_of_organisation(
+            base_ids=source_base_ids, organisation_id=source_organisation_id
+        )
+        _validate_bases_as_part_of_organisation(
+            base_ids=target_base_ids, organisation_id=target_organisation_id
+        )
+
+        # Build all combinations of source and target bases under current agreement
```

Comment:
As per definition in `types.graphql`, the type of the agreement must be specified by the client when creating the agreement. There is no default type.
Currently the only implication of a _unidirectional_ shipment is that only members of the source organisation can create shipments. Why do you think it also affects the way bases are stored in the agreement details table?

Reply:
@pylipp Thanks for clarifying, actually I thought if the client specified a unidirectional shipment, then in the agreement, we only needed the shipment from the source base to the target base, and no other combination.

Reply:
I think I see your point now!
What currently happens with a bidirectional agreement and the source bases [1, 2] and target bases [3, 4]:
- the combinations 1-3, 1-4, 2-3, 2-4 are stored as agreement details
- logically it would be possible to create a shipment from 3 to 1 but the current validation (`_validate_bases_as_part_of_transfer_agreement` in `shipment.py`) would not allow this. So indeed we need to create the combinations 3-1, 4-1, 3-2, 4-2 as well (for bidirectional agreements).

In general then the terms 'source' and 'target' base are a little confusing because the bases of both organisations can be source _and_ target. But I think with the full combinations it's clearer already.

Reply:
@pylipp Thank you for clarifying, you are absolutely right, I was confused at first.

Reply:
Note that I adjusted the implementation of `_validate_bases_as_part_of_transfer_agreement` but did not end up adding all base combinations to the agreement details. This way the fields `sourceBases` and `targetBases` of the `TransferAgreement` type list the source organisation bases and target organisation bases included in the agreement, instead of all bases that might be used as source/target of a shipment.